eni: fix new node not triggering creation of ENI with fix deadlock

wu0407 opened the pull request:
1. Sync the instances network interface information from instances api when a new node is created but not in `instanceManager.instances`. This is fix errors
 in `ResyncInterfacesAndIPs` "Instance not found! Please delete corresponding
ciliumnode if instance has already been deleted" when new node created in cloud provider then add to cluster and before resync triggered.

2. Delete instance from `instanceManager.instances` on node deleted. This will cause `NodeManager.Update()` to invoke `instancesAPIResync` if this instance rejoins the cluster. This ensures that `Node.recalculate()` does not use stale data for instances which rejoin the cluster after their EC2 configuration has changed.

3. Fixes daedlock that report in #21477

Fixes: #20678

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

gandro reviewed (changes requested):
(see comment above)

gandro commented on pkg/ipam/node_manager.go:
```diff
@@ -288,6 +294,16 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
+
 		node.ops = n.instancesAPI.CreateNode(resource, node)
 
 		ctx, cancel := context.WithCancel(context.Background())
```
One last request: Do you mind moving this up and pass `ctx` it into `n.instancesAPI.Resync`? It doesn't affect correctness (since the context won't be cancelled till later), but if we ever change the behavior of `Resync` to spawn any go routines, it will ensure that those are cancelled when the node is removed.

wu0407 replied:
> One last request: Do you mind moving this up and pass `ctx` it into `n.instancesAPI.Resync`? It doesn't affect correctness (since the context won't be cancelled till later), but if we ever change the behavior of `Resync` to spawn any go routines, it will ensure that those are cancelled when the node is removed.

ok

christarazi commented on operator/cmd/cilium_node.go:
```diff
@@ -269,8 +274,18 @@ func syncHandlerConstructor(notFoundHandler func(name string), foundHandler func
+			cn, ok = tombstone.Obj.(*cilium_v2.CiliumNode)
+			if !ok {
+				return fmt.Errorf("tombstone contained object that is not a *cilium_v2.CiliumNode %#v", obj)
+			}
```
I don't think we should be splatting the obj into the error msg. We only care about the type, right?

```suggestion
				return fmt.Errorf("couldn't get object from tombstone %T", obj)
			}
			cn, ok = tombstone.Obj.(*cilium_v2.CiliumNode)
			if !ok {
				return fmt.Errorf("tombstone contained object that is not a *cilium_v2.CiliumNode %T", obj)
			}
```

christarazi reviewed (changes requested):
LGTM besides one comment.

tommyp1ckles commented on pkg/azure/ipam/instances.go:
```diff
@@ -46,6 +46,14 @@ func (m *InstancesManager) CreateNode(obj *v2.CiliumNode, n *ipam.Node) ipam.Nod
+// HasInstance returns whether the instance is in instances
+func (m *InstancesManager) HasInstance(instanceID string) bool {
+	m.mutex.RLock()
+	instanceExist := m.instances.Exists(instanceID)
```
nit: might be a bit cleaner to just return the result of .Exists() and defer the RUnlock()

tommyp1ckles reviewed (approved):
Azure changes LGTM

wu0407 committed:
eni: fix new node not triggering creation of ENI

1. Sync the instances network interface information from instances api when a
new node is created but not in `instanceManager.instances`. This is fix errors
 in `ResyncInterfacesAndIPs` "Instance not found! Please delete corresponding
ciliumnode if instance has already been deleted" when new node created in
cloud provider then add to cluster and before resync triggered.

2. Delete instance from `instanceManager.instances` on node deleted. This will
cause `NodeManager.Update()` to invoke `instancesAPIResync` if this instance
rejoins the cluster. This ensures that `Node.recalculate()` does not use stale
data for instances which rejoin the cluster after their EC2 configuration has changed.

3. Fixes daedlock that report in #21477

Fixes: #20678

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>
